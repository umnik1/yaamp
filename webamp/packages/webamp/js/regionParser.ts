import { parseIni } from "./utils";

export function pointPairs(arr: string[]) {
  const pairedValues = [];
  for (let i = 0; i < arr.length; i += 2) {
    pairedValues.push(`${arr[i]},${arr[i + 1]}`);
  }
  return pairedValues;
}

type RegionData = { [section: string]: string[] };

export default function regionParser(regionStr: string): RegionData {
  const iniData = parseIni(regionStr);
  const data: RegionData = {};
  Object.keys(iniData).forEach((section) => {
    const { numpoints, pointlist } = iniData[section];
    if (!numpoints || !pointlist) {
      return;
    }
    const pointCounts = numpoints.split(/\s*,\s*/).filter((val) => val !== "");
    const points = pointPairs(
      // points can be separated by spaces, or by commas
      pointlist.split(/\s*[, ]\s*/).filter((val) => val !== "")
    );
    let pointIndex = 0;
    const polygons = pointCounts.map((numStr) => {
      const num = Number(numStr);
      if (num < 3) {
        // What use is a polygon with less than three sides?
        pointIndex += num;
        return null;
      }
      const polygon = points.slice(pointIndex, pointIndex + num).join(" ");
      if (!polygon.length) {
        // It's possible that the skin author specified more polygons than provided points.
        return null;
      }
      pointIndex += num;
      return polygon;
    });
    const validPolygons = polygons.filter(
      (polygon) => polygon != null
    ) as string[];
    if (validPolygons.length) {
      data[section] = validPolygons;
    }
  });

  return data;
}
