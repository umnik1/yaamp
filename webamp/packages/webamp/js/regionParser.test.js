import fs from "fs";
import regionParser, { pointPairs } from "./regionParser";

describe("pointPairs", () => {
  it("8", () => {
    expect(
      pointPairs(["1", "0", "275", "0", "275", "115", "1", "115"])
    ).toEqual(["1,0", "275,0", "275,115", "1,115"]);
  });
});

describe("regionParser", () => {
  it("parses the default file as empty", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toEqual({});
  });
  it("parses a complex file", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region1.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
  it("parses a file with section headers but no info", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region_empty_sections.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
  it("parses the EVAunit region.txt", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region_eva.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
  it("parses the iTuned region.txt", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region_ituned.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
  it("parses a region.txt where the points have leading commas", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region_leading_comma.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
  it("parses Satellite M's region.txt", () => {
    const regionTxt = fs.readFileSync(
      "./js/__tests__/fixtures/region_satellite.txt",
      "utf8"
    );
    expect(regionParser(regionTxt)).toMatchSnapshot();
  });
});
