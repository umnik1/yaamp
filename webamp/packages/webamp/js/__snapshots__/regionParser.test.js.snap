// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regionParser parses Satellite M's region.txt 1`] = `
Object {
  "normal": Array [
    "4,0 271,0 271,116 4,116",
    "2,1 273,1 273,116 2,116",
    "1,2 274,2 274,116 1,116",
    "0,4 275,4 275,116 0,116",
  ],
}
`;

exports[`regionParser parses a complex file 1`] = `
Object {
  "equalizer": Array [
    "2,0 9,0 9,2 2,2",
    "9,0 25,0 25,1 9,1",
    "41,0 51,0 51,10 41,10",
    "51,0 81,0 81,1 51,1",
    "82,0 122,0 122,10 82,10",
    "123,0 133,0 133,10 123,10",
    "133,0 163,0 163,1 133,1",
    "1,1 3,1 3,3 1,3",
    "15,1 25,1 25,10 15,10",
    "25,1 51,1 51,9 25,9",
    "57,1 59,1 59,3 57,3",
    "59,1 65,1 65,2 59,2",
    "71,1 81,1 81,10 71,10",
    "81,1 133,1 133,9 81,9",
    "139,1 141,1 141,3 139,3",
    "141,1 147,1 147,2 141,2",
    "152,1 159,1 159,10 152,10",
    "161,1 162,1 162,10 161,10",
    "0,2 2,2 2,9 0,9",
    "7,2 10,2 10,3 7,3",
    "63,2 65,2 65,3 63,3",
    "145,2 148,2 148,3 145,3",
    "159,2 160,2 160,10 159,10",
    "166,2 168,2 168,16 166,16",
    "8,3 10,3 10,8 8,8",
    "56,3 58,3 58,7 56,7",
    "64,3 66,3 66,7 64,7",
    "138,3 140,3 140,7 138,7",
    "146,3 149,3 149,5 146,5",
    "160,3 162,3 162,10 160,10",
    "165,3 172,3 172,13 165,13",
    "172,3 188,3 188,4 172,4",
    "218,3 228,3 228,115 218,115",
    "255,3 260,3 260,12 255,12",
    "264,3 269,3 269,12 264,12",
    "10,4 11,4 11,7 10,7",
    "149,4 150,4 150,7 149,7",
    "164,4 172,4 172,13 164,13",
    "178,4 188,4 188,13 178,13",
    "188,4 269,4 269,12 188,12",
    "271,4 273,4 273,12 271,12",
    "11,5 12,5 12,7 11,7",
    "146,5 148,5 148,7 146,7",
    "150,5 151,5 151,7 150,7",
    "163,5 172,5 172,12 163,12",
    "172,5 173,5 173,11 172,11",
    "269,5 273,5 273,11 269,11",
    "12,6 14,6 14,8 12,8",
    "151,6 173,6 173,10 151,10",
    "2,7 3,7 3,10 2,10",
    "7,7 9,7 9,9 7,9",
    "14,7 18,7 18,11 14,11",
    "57,7 59,7 59,10 57,10",
    "63,7 65,7 65,10 63,10",
    "139,7 141,7 141,10 139,10",
    "145,7 147,7 147,10 145,10",
    "3,8 8,8 8,10 3,10",
    "13,8 51,8 51,9 13,9",
    "59,8 65,8 65,10 59,10",
    "141,8 147,8 147,10 141,10",
    "173,8 174,8 174,10 173,10",
    "0,9 1,9 1,25 0,25",
    "51,9 81,9 81,10 51,10",
    "133,9 174,9 174,10 133,10",
    "4,10 7,10 7,11 4,11",
    "13,10 18,10 18,11 13,11",
    "141,10 143,10 143,11 141,11",
    "150,10 154,10 154,11 150,11",
    "160,10 161,10 161,13 160,13",
    "162,10 173,10 173,11 162,11",
    "5,11 7,11 7,12 5,12",
    "12,11 14,11 14,12 12,12",
    "17,11 19,11 19,13 17,13",
    "142,11 145,11 145,12 142,12",
    "149,11 151,11 151,12 149,12",
    "153,11 155,11 155,13 153,13",
    "159,11 161,11 161,13 159,13",
    "6,12 8,12 8,13 6,13",
    "11,12 13,12 13,13 11,13",
    "19,12 20,12 20,14 19,14",
    "143,12 145,12 145,13 143,13",
    "148,12 150,12 150,13 148,13",
    "172,12 188,12 188,13 172,13",
    "7,13 9,13 9,116 7,116",
    "10,13 12,13 12,14 10,14",
    "20,13 21,13 21,116 20,116",
    "116,13 126,13 126,115 116,115",
    "144,13 148,13 148,16 144,16",
    "148,13 149,13 149,14 148,14",
    "154,13 156,13 156,14 154,14",
    "158,13 160,13 160,14 158,14",
    "165,13 170,13 170,14 165,14",
    "9,14 11,14 11,116 9,116",
    "21,14 22,14 22,116 21,116",
    "24,14 64,14 64,107 24,107",
    "65,14 75,14 75,115 65,115",
    "75,14 105,14 105,15 75,15",
    "137,14 147,14 147,115 137,115",
    "155,14 159,14 159,16 155,16",
    "168,14 177,14 177,115 168,115",
    "190,14 201,14 201,115 190,115",
    "201,14 214,14 214,15 201,15",
    "1,15 11,15 11,104 1,104",
    "14,15 16,15 16,116 14,116",
    "18,15 75,15 75,107 18,107",
    "81,15 83,15 83,17 81,17",
    "83,15 89,15 89,16 83,16",
    "95,15 147,15 147,115 95,115",
    "148,15 150,15 150,17 148,17",
    "153,15 155,15 155,17 153,17",
    "159,15 161,15 161,17 159,17",
    "164,15 166,15 166,17 164,17",
    "177,15 201,15 201,115 177,115",
    "207,15 209,15 209,17 207,17",
    "209,15 215,15 215,16 209,16",
    "16,16 75,16 75,107 16,107",
    "80,16 82,16 82,22 80,22",
    "87,16 90,16 90,17 87,17",
    "150,16 151,16 151,18 150,18",
    "152,16 154,16 154,22 152,22",
    "163,16 166,16 166,17 163,17",
    "167,16 201,16 201,115 167,115",
    "206,16 208,16 208,22 206,22",
    "213,16 216,16 216,17 213,17",
    "11,17 13,17 13,116 11,116",
    "88,17 90,17 90,21 88,21",
    "151,17 154,17 154,18 151,18",
    "160,17 162,17 162,21 160,21",
    "162,17 164,17 164,18 162,18",
    "214,17 216,17 216,23 214,23",
    "205,18 208,18 208,21 205,21",
    "230,18 236,18 236,115 230,115",
    "239,18 245,18 245,115 239,115",
    "247,18 258,18 258,115 247,115",
    "258,18 271,18 271,19 258,19",
    "13,19 75,19 75,107 13,107",
    "204,19 208,19 208,21 204,21",
    "228,19 258,19 258,115 228,115",
    "264,19 266,19 266,21 264,21",
    "266,19 272,19 272,20 266,20",
    "78,20 82,20 82,21 78,21",
    "90,20 92,20 92,21 90,21",
    "150,20 154,20 154,21 150,21",
    "162,20 164,20 164,21 162,21",
    "202,20 204,20 204,22 202,22",
    "263,20 265,20 265,26 263,26",
    "270,20 273,20 273,21 270,21",
    "76,21 78,21 78,23 76,23",
    "78,21 79,21 79,22 78,22",
    "82,21 83,21 83,23 82,23",
    "87,21 89,21 89,23 87,23",
    "91,21 94,21 94,22 91,22",
    "148,21 150,21 150,23 148,23",
    "150,21 151,21 151,22 150,22",
    "154,21 155,21 155,23 154,23",
    "159,21 161,21 161,23 159,23",
    "163,21 166,21 166,22 163,22",
    "201,21 203,21 203,23 201,23",
    "208,21 209,21 209,24 208,24",
    "213,21 258,21 258,23 213,23",
    "271,21 273,21 273,27 271,27",
    "75,22 78,22 78,23 75,23",
    "81,22 89,22 89,23 81,23",
    "92,22 150,22 150,23 92,23",
    "153,22 161,22 161,23 153,23",
    "164,22 203,22 203,23 164,23",
    "207,22 258,22 258,23 207,23",
    "262,22 265,22 265,25 262,25",
    "75,23 76,23 76,115 75,115",
    "83,23 87,23 87,24 83,24",
    "94,23 148,23 148,115 94,115",
    "155,23 159,23 159,24 155,24",
    "166,23 202,23 202,115 166,115",
    "209,23 212,23 212,25 209,25",
    "212,23 214,23 214,24 212,24",
    "215,23 258,23 258,115 215,115",
    "261,23 265,23 265,25 261,25",
    "76,24 77,24 77,115 76,115",
    "82,24 84,24 84,25 82,25",
    "86,24 88,24 88,25 86,25",
    "93,24 149,24 149,115 93,115",
    "154,24 156,24 156,25 154,25",
    "158,24 160,24 160,25 158,25",
    "165,24 203,24 203,115 165,115",
    "259,24 261,24 261,26 259,26",
    "77,25 78,25 78,115 77,115",
    "81,25 83,25 83,27 81,27",
    "87,25 89,25 89,27 87,27",
    "92,25 150,25 150,115 92,115",
    "153,25 155,25 155,27 153,27",
    "159,25 161,25 161,27 159,27",
    "164,25 204,25 204,115 164,115",
    "209,25 211,25 211,26 209,26",
    "258,25 260,25 260,27 258,27",
    "265,25 266,25 266,28 265,28",
    "270,25 273,25 273,27 270,27",
    "78,26 79,26 79,115 78,115",
    "91,26 151,26 151,115 91,115",
    "163,26 205,26 205,115 163,115",
    "208,26 210,26 210,27 208,27",
    "264,26 273,26 273,27 264,27",
    "79,27 82,27 82,115 79,115",
    "88,27 89,27 89,115 88,115",
    "90,27 154,27 154,115 90,115",
    "160,27 161,27 161,115 160,115",
    "162,27 206,27 206,115 162,115",
    "207,27 209,27 209,115 207,115",
    "258,27 259,27 259,109 258,109",
    "266,27 269,27 269,29 266,29",
    "269,27 271,27 271,28 269,28",
    "272,27 273,27 273,42 272,42",
    "89,28 154,28 154,115 89,115",
    "161,28 209,28 209,115 161,115",
    "259,28 260,28 260,106 259,106",
    "82,29 258,29 258,115 82,115",
    "260,29 261,29 261,105 260,105",
    "266,29 268,29 268,30 266,30",
    "261,30 262,30 262,104 261,104",
    "265,30 267,30 267,31 265,31",
    "262,31 263,31 263,103 262,103",
    "264,31 266,31 266,103 264,103",
    "263,32 266,32 266,102 263,102",
    "266,33 273,33 273,42 266,42",
    "266,42 272,42 272,101 266,101",
    "272,91 273,91 273,114 272,114",
    "265,103 267,103 267,104 265,104",
    "2,104 36,104 36,116 2,116",
    "266,104 268,104 268,108 266,108",
    "268,105 269,105 269,108 268,108",
    "0,106 36,106 36,116 0,116",
    "265,106 271,106 271,108 265,108",
    "42,107 44,107 44,109 42,109",
    "44,107 50,107 50,108 44,108",
    "56,107 66,107 66,116 56,116",
    "259,107 260,107 260,110 259,110",
    "264,107 266,107 266,109 264,109",
    "271,107 273,107 273,114 271,114",
    "48,108 50,108 50,109 48,109",
    "260,108 261,108 261,110 260,110",
    "263,108 265,108 265,114 263,114",
    "270,108 273,108 273,109 270,109",
    "41,109 43,109 43,113 41,113",
    "49,109 51,109 51,113 49,113",
    "261,109 265,109 265,111 261,111",
    "262,111 265,111 265,112 262,112",
    "42,113 44,113 44,116 42,116",
    "48,113 50,113 50,116 48,116",
    "265,113 266,113 266,116 265,116",
    "270,113 272,113 272,115 270,115",
    "44,114 50,114 50,116 44,116",
    "264,114 271,114 271,116 264,116",
    "36,115 66,115 66,116 36,116",
    "248,115 271,115 271,116 248,116",
  ],
  "equalizerws": Array [
    "4,3 29,3 29,13 4,13",
    "48,3 60,3 60,13 48,13",
    "158,3 164,3 164,13 158,13",
    "209,3 249,3 249,13 209,13",
    "255,3 260,3 260,12 255,12",
    "265,3 270,3 270,12 265,12",
    "29,4 271,4 271,12 29,12",
    "273,4 275,4 275,12 273,12",
    "271,5 275,5 275,11 271,11",
  ],
  "normal": Array [
    "8,3 18,3 18,27 8,27",
    "85,3 95,3 95,111 85,111",
    "95,3 125,3 125,4 95,4",
    "152,3 162,3 162,112 152,112",
    "162,3 192,3 192,4 162,4",
    "245,3 250,3 250,111 245,111",
    "255,3 260,3 260,112 255,112",
    "265,3 270,3 270,15 265,15",
    "2,4 4,4 4,12 2,12",
    "6,4 95,4 95,12 6,12",
    "101,4 103,4 103,6 101,6",
    "103,4 109,4 109,5 103,5",
    "115,4 162,4 162,111 115,111",
    "168,4 170,4 170,6 168,6",
    "170,4 176,4 176,5 170,5",
    "182,4 267,4 267,98 182,98",
    "270,4 271,4 271,12 270,12",
    "273,4 275,4 275,12 273,12",
    "4,5 95,5 95,11 4,11",
    "100,5 102,5 102,11 100,11",
    "107,5 110,5 110,6 107,6",
    "174,5 176,5 176,6 174,6",
    "271,5 275,5 275,11 271,11",
    "108,6 110,6 110,10 108,10",
    "167,6 169,6 169,10 167,10",
    "175,6 177,6 177,10 175,10",
    "98,9 102,9 102,10 98,10",
    "110,9 112,9 112,10 110,10",
    "96,10 98,10 98,12 96,12",
    "98,10 99,10 99,11 98,11",
    "102,10 103,10 103,12 102,12",
    "107,10 109,10 109,12 107,12",
    "111,10 114,10 114,11 111,11",
    "168,10 170,10 170,112 168,112",
    "174,10 176,10 176,111 174,111",
    "95,11 98,11 98,12 95,12",
    "101,11 109,11 109,12 101,12",
    "112,11 162,11 162,12 112,12",
    "170,11 176,11 176,111 170,111",
    "18,12 96,12 96,111 18,111",
    "103,12 107,12 107,13 103,13",
    "114,12 260,12 260,111 114,111",
    "96,13 97,13 97,112 96,112",
    "102,13 104,13 104,14 102,14",
    "106,13 108,13 108,14 106,14",
    "113,13 260,13 260,111 113,111",
    "270,13 273,13 273,15 270,15",
    "97,14 98,14 98,112 97,112",
    "101,14 103,14 103,16 101,16",
    "107,14 109,14 109,16 107,16",
    "112,14 260,14 260,111 112,111",
    "273,14 274,14 274,22 273,22",
    "98,15 99,15 99,112 98,112",
    "111,15 260,15 260,111 111,111",
    "267,15 268,15 268,16 267,16",
    "272,15 275,15 275,16 272,16",
    "99,16 102,16 102,111 99,111",
    "108,16 109,16 109,111 108,111",
    "110,16 260,16 260,111 110,111",
    "274,16 275,16 275,21 274,21",
    "109,17 260,17 260,111 109,111",
    "102,18 260,18 260,111 102,111",
    "267,20 268,20 268,100 267,100",
    "272,20 274,20 274,22 272,22",
    "268,21 272,21 272,97 268,97",
    "272,22 273,22 273,23 272,23",
    "272,24 273,24 273,97 272,97",
    "273,25 274,25 274,97 273,97",
    "274,26 275,26 275,40 274,40",
    "9,27 260,27 260,111 9,111",
    "8,28 260,28 260,111 8,111",
    "274,43 275,43 275,53 274,53",
    "2,59 4,59 4,67 2,67",
    "6,59 275,59 275,67 6,67",
    "4,60 275,60 275,66 4,66",
    "274,67 275,67 275,69 274,69",
    "7,69 274,69 274,71 7,71",
    "6,72 260,72 260,111 6,111",
    "0,73 2,73 2,81 0,81",
    "4,73 274,73 274,81 4,81",
    "2,74 274,74 274,80 2,80",
    "274,87 275,87 275,110 274,110",
    "260,98 263,98 263,101 260,101",
    "263,98 265,98 265,99 263,99",
    "266,98 268,98 268,99 266,99",
    "263,99 264,99 264,100 263,100",
    "268,99 269,99 269,104 268,104",
    "269,100 270,100 270,104 269,104",
    "260,101 262,101 262,102 260,102",
    "270,101 271,101 271,104 270,104",
    "260,102 261,102 261,105 260,105",
    "267,102 273,102 273,104 267,104",
    "0,103 2,103 2,111 0,111",
    "4,103 260,103 260,111 4,111",
    "261,103 262,103 262,105 261,105",
    "266,103 268,103 268,105 266,105",
    "273,103 275,103 275,110 273,110",
    "2,104 260,104 260,110 2,110",
    "262,104 263,104 263,106 262,106",
    "265,104 267,104 267,110 265,110",
    "272,104 275,104 275,105 272,105",
    "263,105 267,105 267,107 263,107",
    "264,107 267,107 267,108 264,108",
    "267,109 268,109 268,112 267,112",
    "272,109 274,109 274,111 272,111",
    "266,110 273,110 273,112 266,112",
    "6,111 16,111 16,112 6,112",
    "21,111 31,111 31,112 21,112",
    "44,111 54,111 54,112 44,112",
    "67,111 77,111 77,112 67,112",
    "90,111 100,111 100,112 90,112",
    "112,111 122,111 122,112 112,112",
    "133,111 173,111 173,112 133,112",
    "249,111 273,111 273,112 249,112",
  ],
  "windowshade": Array [
    "8,3 18,3 18,13 8,13",
    "78,3 192,3 192,13 78,13",
    "245,3 250,3 250,12 245,12",
    "255,3 260,3 260,12 255,12",
    "265,3 270,3 270,12 265,12",
    "2,4 4,4 4,12 2,12",
    "6,4 271,4 271,12 6,12",
    "273,4 275,4 275,12 273,12",
    "4,5 275,5 275,11 4,11",
  ],
}
`;

exports[`regionParser parses a file with section headers but no info 1`] = `
Object {
  "equalizer": Array [
    "0,0 0,115 20,20 274,115 274,0",
  ],
  "normal": Array [
    "1,0 275,0 275,115 1,115",
  ],
}
`;

exports[`regionParser parses a region.txt where the points have leading commas 1`] = `
Object {
  "equalizer": Array [
    "5,0 3,1 2,2 1,3 0,5 0,116 275,116 275,6 274,3 273,2 272,1 270,0",
  ],
  "equalizerws": Array [
    "5,0 3,1 2,2 1,3 0,5 0,14 275,14 275,6 274,3 273,2 272,1 270,0",
  ],
  "normal": Array [
    "5,0 3,1 2,2 1,3 0,5 0,116 275,116 275,6 274,3 273,2 272,1 270,0",
  ],
  "windowshade": Array [
    "5,0 3,1 2,2 1,3 0,5 0,14 275,14 275,6 274,3 273,2 272,1 270,0",
  ],
}
`;

exports[`regionParser parses the EVAunit region.txt 1`] = `
Object {
  "equalizer": Array [
    "4,0 4,3 2,3 0,5 0,116 275,116 275,5 273,3 270,3 270,0 237,0 237,3 183,3 183,0 129,0 129,3 17,3 17,0",
  ],
  "normal": Array [
    "9,0 9,3 2,3 0,5 0,104 2,107 4,107 4,116 17,116 17,107 129,107 129,116 183,116 183,107 237,107 237,116 270,116 270,107 272,107 275,104 275,5 273,3 271,3 272,4 266,4 266,3 261,3 262,4 256,4 256,3 251,3 252,4 246,4 246,3 243,3 243,0",
  ],
  "windowshade": Array [
    "9,0 9,3 2,3 0,5 0,14 275,14 275,5 273,3 271,3 271,4 266,4 266,3 261,3 261,4 256,4 256,3 251,3 251,4 246,4 246,3 243,3 243,0 157,0 157,3 130,3 130,0",
  ],
}
`;

exports[`regionParser parses the iTuned region.txt 1`] = `
Object {
  "equalizer": Array [
    "5,0 269,0 269,116 5,116",
    "3,1 5,1 5,115 3,115",
    "2,2 4,2 4,114 2,114",
    "1,3 4,3 4,113 1,113",
    "0,5 5,5 5,111 0,111",
    "267,1 271,1 271,115 267,115",
    "267,2 272,2 272,114 267,114",
    "267,3 273,3 273,113 267,113",
    "267,4 274,4 274,112 267,112",
    "267,6 275,6 275,111 267,111",
  ],
  "normal": Array [
    "5,0 269,0 269,116 5,116",
    "3,1 5,1 5,115 3,115",
    "2,2 4,2 4,114 2,114",
    "1,3 4,3 4,113 1,113",
    "0,5 5,5 5,111 0,111",
    "267,1 271,1 271,115 267,115",
    "267,2 272,2 272,114 267,114",
    "267,3 273,3 273,113 267,113",
    "267,4 274,4 274,112 267,112",
    "267,6 275,6 275,111 267,111",
  ],
}
`;
